rpcs3/Emu/NP/np_handler.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include "Emu/Memory/vm_ptr.h"
+#include "Emu/Cell/Modules/sceNp.h"
+#include "Emu/Cell/Modules/sceNp2.h"
+
+#include <queue>
+
+enum Nph_Internet_Status : s32
+{
+	Nph_Internet_Disabled,
+	Nph_Internet_Enabled,
+};
+
+enum Nph_PSN_Status : s32
+{
+	Nph_PSN_Disabled,
+	Nph_PSN_Fake,
+};
+
+class np_handler
+{
+public:
+	np_handler();
+
+	s32 get_psn_status() const;
+	s32 get_net_status() const;
+
+	const std::string& get_ip() const;
+	u32 get_dns() const;
+
+	const SceNpId& get_npid() const;
+	const SceNpOnlineId& get_online_id() const;
+	const SceNpOnlineName& get_online_name() const;
+	const SceNpAvatarUrl& get_avatar_url() const;
+
+	// DNS hooking functions
+	void add_dns_spy(u32 sock);
+	void remove_dns_spy(u32 sock);
+	bool is_dns(u32 sock) const;
+	bool is_dns_queue(u32 sock) const;
+	std::vector<u8> get_dns_packet(u32 sock);
+	s32 analyze_dns_packet(s32 s, const u8* buf, u32 len);
+
+	void operator()();
+
+	void init_NP(u32 poolsize, vm::ptr<void> poolptr);
+	void terminate_NP();
+
+	bool is_netctl_init     = false;
+	bool is_NP_init         = false;
+	bool is_NP_Lookup_init  = false;
+	bool is_NP_Score_init   = false;
+	bool is_NP2_init        = false;
+	bool is_NP2_Match2_init = false;
+	bool is_NP_Auth_init    = false;
+
+	// Score related
+	struct ScoreCtx
+	{
+		SceNpCommunicationId communicationId;
+		SceNpCommunicationPassphrase passphrase;
+	};
+	s32 create_score_context(vm::cptr<SceNpCommunicationId> communicationId, vm::cptr<SceNpCommunicationPassphrase> passphrase);
+
+	// Match2 related
+	struct Match2Ctx
+	{
+		SceNpCommunicationId communicationId;
+		SceNpCommunicationPassphrase passphrase;
+	};
+	u16 create_match2_context(vm::cptr<SceNpCommunicationId> communicationId, vm::cptr<SceNpCommunicationPassphrase> passphrase);	
+
+	static constexpr auto thread_name = "NP Handler Thread"sv;
+
+protected:
+	bool is_connected  = false;
+	bool is_psn_active = false;
+
+	// Net infos
+	std::string cur_ip{};
+	u32 cur_addr = 0;
+	u32 dns      = 0x08080808;
+
+	// User infos
+	SceNpId npid{};
+	SceNpOnlineName online_name{};
+	SceNpAvatarUrl avatar_url{};
+
+	// DNS related
+	std::map<s32, std::queue<std::vector<u8>>> dns_spylist{};
+	std::map<std::string, u32> switch_map{};
+
+	// Memory pool for sceNp/sceNp2
+	vm::ptr<void> mpool{};
+	u32 mpool_size  = 0;
+	u32 mpool_avail = 0;
+	std::map<u32, u32> mpool_allocs{}; // offset/size
+	vm::addr_t allocate(u32 size);
+
+	// Contexes
+	std::map<s32, ScoreCtx> score_ctx_map{};
+	s32 score_ctx_cnt = 1;
```

Comment:
Without looking at the code, only checking the naming convention here - won't it be redundant with `score_ctx_map.size() + 1`?

Reply:
I am wondering why IDM was not being used here instead.